src/main/java/com/baidu/hugegraph/base/RetryManager.java

```diff
@@ -85,18 +88,29 @@ public RetryManager(String url, String graph,
         return r;
     }
 
-    public void submit(Runnable task) {
+    public synchronized void submit(Runnable task) {
         this.futures.add(this.pool.submit(task));
     }
 
     public void awaitTasks() {
-        for (Future<?> future : this.futures) {
-            try {
-                future.get();
-            } catch (InterruptedException | ExecutionException e) {
-                e.printStackTrace();
+        int size = 0;
+        int offset;
+        do {
+            offset = size;
+            size = this.futures.size();
+            for (int i = offset; i < size; i++) {
+                Future<?> future = this.futures.get(i);
+                try {
+                    future.get();
+                } catch (InterruptedException | ExecutionException e) {
+                    e.printStackTrace();
+                }
             }
-        }
+        } while (size < this.futures.size());
+        this.clearFutures();
+    }
+
+    private synchronized void clearFutures() {
```

Comment:
remove clearFutures() and change futures to Collections.synchronizedList
also remove synchronized at line 91